### ISSUE TITLE
ncurses incompatibility with the latest conda and bioconda/conda-forge

### ISSUE THREAD

**Issue by chapmanb** [bioconda/bioconda-recipes#6219]
There appear to be ncurses incompability issues with the latest conda release from Miniconda (4.3.27) related to a move to gcc 7.x. conda now depends on libedit, which pins to curses 6.0:
```
$ conda info libedit
Fetching package metadata ...........

libedit 3.1 heed3624_0
----------------------
file name   : libedit-3.1-heed3624_0.tar.bz2
name        : libedit
version     : 3.1
build string: heed3624_0
build number: 0
channel     : defaults
size        : 171 KB
arch        : None
license     : 3-Clause BSD
license_family: BSD
md5         : 91ce7622c3bc0a51cb3b897749f1aeaf
noarch      : None
platform    : None
subdir      : linux-64
timestamp   : 1505666480132
url         : https://repo.continuum.io/pkgs/main/linux-64/libedit-3.1-heed3624_0.tar.bz2
dependencies:
    libgcc-ng >=7.2.0
    ncurses 6.0.*
```
This causes incompatibility issues with bioconda/conda-forge tools that pin to curses 5.9 (see chapmanb/bcbio-nextgen#2090). As a workaround you can do:
```
conda uninstall --force libedit
```
and conda appears to function okay.

I'm not clued in enough about the migration efforts to know if this is known and already being tackled but I wanted to post the work around and start a discussion with @bioconda/core around better ways to resolve it.

**Comment by johanneskoester** [comment]
@jakirkham I tried to find a related issue on conda-forge, but no luck so far. Do you guys have a strategy?

**Comment by johanneskoester** [comment]
I think this problem only occurs in the root environment. When you create new environments, conda is able to handle this, by simply using the older stuff.

**Comment by lwintermelon** [comment]
@joachimwolff it is strange that create new environments works for linux, but not for mac...

**Comment by jakirkham** [comment]
To answer your question, @johanneskoester, not really no. Am not aware of any such effort.

More generally, the real question is what is pulling in `libedit`? Also is this an issue that is occurring with an existing environment? What channels are involved, how (e.g. via command like or in config), and in which order?

For those that don't know, `libedit` is a BSD alternative to `readline`. Things like `python` and the R stack should all be using `readline` in `conda-forge`. So I'm unclear on how `libedit` is getting involved.

**Comment by chapmanb** [comment]
Thanks for helping look at this. libedit gets installed as part of a standard Miniconda install:
```
[:~/tmp/bcbio/install] $ bash Miniconda2-latest-Linux-x86_64.sh -b -p venv
PREFIX=/home/chapmanb/tmp/bcbio/install/venv
installing: python-2.7.13-heccc3f1_16 ...
Python 2.7.13 :: Anaconda, Inc.
[...]
installing: libedit-3.1-heed3624_0 ...
```
so it's part of the default install. Looking through the `conda-meta` json files it appears to be a dependency of sqlite-3.20.1-h6d8b0f3_1:
```
  "depends": [
    "libedit",
    "libgcc-ng >=7.2.0"
```
which is a dependency of python-2.7.13-heccc3f1_16:
```
  "depends": [
    "libffi 3.2.*",
    "libgcc-ng >=7.2.0",
    "libstdcxx-ng >=7.2.0",
    "ncurses 6.0.*",
    "openssl 1.0.*",
    "readline 7.*",
    "sqlite >=3.20.1,<4.0a0",
    "tk 8.6.*",
    "zlib >=1.2.11,<1.3.0a0"
```
which also pins to ncurses 6.0.* so might indicate the potential for additional conflicts in the future.

I don't know enough about the ncurses pinning in conda-forge versus standard packages to understand the best approach going forward but was hoping to raise this issue to discuss how to remain compatible. Thanks again for the help.

**Comment by jakirkham** [comment]
I see. So in `root`. Well one can uninstall `libedit` and then update everything to use `conda-forge`. That should fix the issue.

**Comment by johanneskoester** [comment]
Sure. But that would only be a workaround. It would be nicer to have it working out of the box, woudn't it?

**Comment by jakirkham** [comment]
For now what we have is a workaround. Another one would be to use an older Miniconda and upgrade to `conda-forge` packages. It's better than nothing no?

A real fix would be to create our own conda-forge installer. We can't keep putting ourselves in situations where `defaults` makes some decisions about what goes in the installer that then cause us issues. The way we fix that is to make our own installer that includes what we want and how we want it to start.

There is one blocker that I know of with this approach. We need to be able to disambiguate `defaults` and `conda-forge` when included. This comes down to issue ( https://github.com/conda/constructor/issues/42 ).

**Comment by johanneskoester** [comment]
I totally agree with you.

What I find disappointing is that Continuum first expresses that they fully support conda-forge and then they "break" it without notifying or discussing with the community in advance.

**Comment by jakirkham** [comment]
So they did [provide previews]( https://gitter.im/conda-forge/conda-forge.github.io?at=59c514c832fc8b7e402c7a2b ) on a Friday and were planning to release on the following Monday. A couple of people had time to try it. I don't recall if anyone found the `libedit` change or the problems it could cause at the time. However I doubt it will go away as the motivation was to switch away from GPL'd packages in Anaconda's installer.

I suppose another workaround we could try is adding `libedit` to `conda-forge` with a pinning that matches `conda-forge`'s. That way `libedit` would get updated to the `conda-forge` copy. Thoughts?

**Comment by chapmanb** [comment]
Thanks for this helpful discussion and background. It sounds like the major breaker is the ncurses version pinning difference, and libedit happens to expose it because it gets installed by base and then remains as a problem when updating other packages to the conda-forge versions. I like the idea of shadowing with a pin-matching libevent which seems cleaner and easier than the other approaches. Out of curiosity, is synchronizing ncurses version pinning a major undertaking? I don't have a good understanding of the approach to doing this, just interested in learning more about how everything works.

**Comment by jakirkham** [comment]
Adding this as a relevant side note. Would also point out that we have asked that `conda` track requested packages independently from unrequested packages in issue ( https://github.com/conda/conda/issues/3968 ). We could view this situation as one where nothing has been requested, in which case removing `libedit` and/or breaking its pinning should be ok. Though we might need to give `conda` more information like we actually want `conda` and maybe `python` to be counted as requested dependencies. It looks like this will be addressed somehow in `conda` 4.4, but have not looked at the implementation personally.

**Comment by jakirkham** [comment]
> I like the idea of shadowing with a pin-matching libevent which seems cleaner and easier than the other approaches.

Guessing you meant `libedit`. If that's ok with others, that seems like the easiest path forward near term.

> Out of curiosity, is synchronizing ncurses version pinning a major undertaking? I don't have a good understanding of the approach to doing this, just interested in learning more about how everything works.

So `ncurses` and `readline`/`libedit` form the backbone of any terminal level interactivity (think using REPLs). I don't think lots of things would be rebuilt off the top of my head, but breakage in this area would be painful and tricky to fix. We might also learn some things like we need a newer version of `readline` or everything using `readline` needs to be rebuilt (even if it wasn't directly linked to `ncurses`). As I'm kind of swamped with other work ATM, may not be able to take this on right now.

**Comment by chapmanb** [comment]
Whoops sorry, I did mean libedit. Understood about the potential impact of the changes, thanks for the explanation and ideas here, much appreciated.

**Comment by jakirkham** [comment]
No worries. Of course. At some point I'm sure we will want to switch to the new `ncurses`, but will want us to do that carefully ( after all many of us are terminal bound :) ). In the interim, we have an issue at conda-forge now to discuss adding `libedit`. ( https://github.com/conda-forge/staged-recipes/issues/4209 ) Please feel free to chime in.

**Comment by mingwandroid** [comment]
> So they did provide previews on a Friday and were planning to release on the following Monday

This is incorrect, we provided previews many many weeks ahead of the release, I found this reference:

https://groups.google.com/a/continuum.io/forum/#!topic/anaconda/EoJj6rGDYgI

I've also been informally asking people to help test Miniconda since SciPy.

**Comment by mingwandroid** [comment]
> A real fix would be to create our own conda-forge installer

A better fix is to update the [very old](https://www.gnu.org/software/ncurses/) ncurses version. The `Anaconda Distribution` cannot provide such old software in good conscience and at each major release we make a concerted effort to update to the latest stable versions of all packages. Hopefully for the next major release (or the one after) those version updates can be routed through conda-forge instead of done in our forks.

This release was particularly stressful since we needed to change so much stuff (basically everything) and I apologise that we haven't had a lot of time to work with getting our changes pushed as PRs, been good at communicating version changes or helping where we can with buildouts.

But please, let's not take steps away from compatibility when improving that was one of the most important goals for AD5 and so much good progress has been made.

**Comment by johanneskoester** [comment]
@mingwandroid perhaps there could be an official communication channel between Continuum and the communities so that we all get notifications and do not have to actively search. For example, a github continuing issue that we could all subscribe to which announces breaking changes (including a test period) and allows us to test?

**Comment by mingwandroid** [comment]
For me, the ideal situation is for our forks to contain absolutely no changes from conda-forge. In some cases it will not be possible, but for most things I believe it should be.

> perhaps there could be an official communication channel

Sure! Suggestions on interop. improvements (both technical and non-technical) are *very* welcome. Ping @ocefpaf, @teoliphant ,@csoja, @msarahan, @nehaljwani and @jjhelmus.

**Comment by jakirkham** [comment]
> This is incorrect, we provided previews many many weeks ahead of the release, I found this reference:
>
> https://groups.google.com/a/continuum.io/forum/#!topic/anaconda/EoJj6rGDYgI
>
> I've also been informally asking people to help test Miniconda since SciPy.

Maybe, but that was not what we were [told]( https://gitter.im/conda-forge/conda-forge.github.io?at=59b87e10210ac26920243ab9 ). I think the advice was to use the new Anaconda at that time and I don't know if anyone tried that or not.

**Comment by mingwandroid** [comment]
That statement from me carried an implicit "use the Anaconda previews instead".

**Comment by johanneskoester** [comment]
@mingwandroid I totally agree with you, keeping everything in sync is the best solution. Looking forward to that! I guess somebody from Continuum should open that tracking issue? Maybe the issue text should say that it should not be used for discussion, in order to make the important stuff more visible to everybody passing by.

**Comment by jjhelmus** [comment]
An update to the conda-forge ncurses pinning has been proposed in conda-forge/conda-forge.github.io#461.  I added a list of the `conda-forge` packages that I believe would need to be rebuilt to that issue.  I'll submit a PR to the [ncurses-feedstock](https://github.com/conda-forge/ncurses-feedstock) later tonight.  

@jakirkham It looks like you and I are maintainers on `ncurses`.  Would you be available to review a PR?

**Comment by jakirkham** [comment]
Not for a couple weeks. This is why I was suggesting adding `libedit` as a near term workaround.

**Comment by mingwandroid** [comment]
`libedit` has no bearing on this issue, it's just a version incompatibility in `ncurses`. Probably it's been seen as important and that's how it was reported initially, but Python depends on `ncurses` anyway. In other words, building `libedit` with `ncurses` pinned to `5.9` will not solve this incompatibility.

**Comment by jakirkham** [comment]
From the issues that I have seen, it seems that `ncurses` fails to downgrade due to being pinned by `libedit` even when all other packages being installed or changed are from say `conda-forge` and do use the older version of `ncurses`. It's true that this won't help with new packages from `defaults` that `conda-forge` doesn't have in some form. Though for anything `conda-forge` already has like `r`, shadowing `libedit` with an older `ncurses` pinning should work.

**Comment by mingwandroid** [comment]
https://github.com/AnacondaRecipes/python-feedstock/blob/master/recipe/meta.yaml#L89

I think you'll be playing whack-a-pinning here. The only correct fix is to modernise. Is there anything I can do to help out with this? I'm familiar with `ncurses` and can help review.

**Comment by mingwandroid** [comment]
OK, I see, yes. Having no such package at all probably means it gets stuck.

I could possibly build new Minicondas with `sqlite` not depending upon `libedit`, but I can't do it until early next week and I also need to get permission to do that.

**Comment by johanneskoester** [comment]
@mingwandroid I totally agree with you, keeping everything in sync is the best solution. Looking forward to that! I guess somebody from Continuum should open that tracking issue? Maybe the issue text should say that it should not be used for discussion, in order to make the important stuff more visible to everybody passing by.

**Comment by jakirkham** [comment]
FWIW I have no issues going ahead with an `ncurses` update. Just have had very little time to help.

Part of the reason I had suggested the `libedit` workaround beyond buying us some time is that I suspect it will be less complicated and thus less likely to run into issues part way. Not to mention it does not preclude updating `ncurses`. Just provides end users with some way to install things while the `ncurses` update is explored.

In any event, `libedit` has been added to `conda-forge`. Linux builds are already available. macOS builds are queued. Expect macOS will at least be available by the end of the weekend if not sooner.
